File: CIS/dp_prog.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import numpy as np
from scipy.signal import find_peaks

def dp_prog(lp_res,avg,delta):

    assert not any(np.isnan(lp_res)), 'First argument can not have NaN in its element'
    
    lamb = 1
    lmin = np.floor((1-delta)*avg).astype(np.int)
    lmax = np.floor((1+delta)*avg).astype(np.int)
    cost_sig = np.zeros((len(lp_res), ))
    loc_sig = np.zeros((len(lp_res), ))
    cost = np.zeros((lmax-lmin+1, ))
    for i in range(lmax+1, len(lp_res)):
        for j in range(lmin, lmax):
            cost[j-lmin] = cost_sig[i-j]-lamb*lp_res[i-j]
        l1 = np.argwhere(cost == np.min(cost))
        loc_sig[i] = i-lmin-l1[0]+1
        cost_sig[i] = np.min(cost)

    k = 0
    last_sig = np.array([cost_sig[np.max(np.clip(np.arange(len(cost_sig)-2*lmax,
                                                 len(cost_sig)), 0, np.inf)).astype(np.int)]])
    l12 = np.argwhere(last_sig == np.min(last_sig))[0]
    last_loc = len(cost_sig)-2*lmax+l12[0]-1
    indices = np.array([last_loc], dtype=np.int)
    while k == 0:
        if loc_sig[indices[0]]<lmax+20000:
            break
        indices = np.append(indices, loc_sig[indices[0]].astype(np.int))

    gci = np.zeros((len(lp_res),))
    gci[indices] = 1
    s2sum = np.zeros((2*np.int(avg/2)+1, ))

    for i in range(-np.int(avg/2),np.int(avg/2)):
        ids = indices+i
        ids = ids[np.argwhere(ids<len(lp_res))]
        s2sum[i+np.int(avg/2)] = np.mean(lp_res[ids])

    i1 = np.round(np.linspace(0, 1, 6)*len(s2sum)); i1[0] = 1
    pksloc, _ = find_peaks(s2sum)
    pksval = s2sum[pksloc]
    indr = np.argwhere((pksloc>=i1[3]) & (pksloc<=i1[5])).reshape(-1)
    indl = np.argwhere((pksloc>=i1[0]) & (pksloc<=i1[2])).reshape(-1)

    if len(indl)==0:
        v = np.max(pksval[indr])
        id = np.argmax(pksval[indr])
        s2loc = pksloc[indr[id]]-np.round(avg/2).astype(np.int)
    elif len(indr)==0:
        v = np.max(pksval[indl])
        id = np.argmax(pksval[indl])
        s2loc = pksloc[indl[id]]-np.round(avg/2).astype(np.int)
    else:
        if np.max(pksval[indr])>np.max(pksval[indl]):
            v = np.max(pksval[indr])
            id = np.argmax(pksval[indr])
            s2loc = pksloc[indr[id]]-np.round(avg/2).astype(np.int)
        else:
            v = np.max(pksval[indl])
            id = np.argmax(pksval[indl])
            s2loc = pksloc[indl[id]]-np.round(avg/2)

    if s2loc>0:
        indices_s1 = indices; indices_s2 = indices+s2loc
        indices_s2 = indices_s2[indices_s2>len(lp_res)]
        indices_s2 = indices_s2[indices_s2<0]
    else:
        indices_s2 = indices; indices_s1 = indices+s2loc
        indices_s1 = indices_s1[indices_s1<len(lp_res)]
        indices_s1 = indices_s1[indices_s1>0]

    
    return gci, cost_sig, loc_sig, indices_s1, indices_s2

File: CIS/matlab_scripts/dp_prog.m
function [gci, cost_sig, loc_sig, indices_s1 indices_s2]=dp_prog(lp_res,avg,delta)

if ~isempty(find(isnan(lp_res)==1))
    error('First argument can not have NaN in its element');
end

lamb=1;
lmin=floor((1-delta)*avg);
lmax=floor((1+delta)*avg);
cost_sig=zeros(1,numel(lp_res));
loc_sig=zeros(1,numel(lp_res));
for i=lmax+1:numel(lp_res)
    for j=lmin:lmax
        cost(j-lmin+1)=cost_sig(i-j)-lamb*lp_res(i-j);
    end
    l1=find(cost==min(cost));
    loc_sig(i)=i-lmin-l1(1)+1;
%     if(loc_sig(loc_sig(i))>0)
%     if(lp_res(loc_sig(i))<.3*lp_res(loc_sig(loc_sig(i))))
%         loc_sig(i)=loc_sig(loc_sig(i));
%     end
%     end
    cost_sig(i)=min(cost);
end
k=0;
last_sig=cost_sig(max(1,numel(cost_sig)-2*lmax):numel(cost_sig));
l12=find(last_sig==min(last_sig));
last_loc=numel(cost_sig)-2*lmax+l12(1)-1;
indices=[last_loc];
while(k==0)
    if loc_sig(indices(1))<lmax+20
        break;
    end
    indices=[loc_sig(indices(1))  indices];
end
gci=zeros(1,numel(lp_res));
gci(indices)=1;


%%%%%%%%%%%
for i=-round(avg/2):1:round(avg/2)

    ids=indices+i;
    ids(find(ids>length(lp_res)))=[];
    ids(find(ids>length(lp_res)))=[];
    s2sum(i+round(avg/2)+1)=mean(lp_res(ids));
    
end
i1=round([0:.2:1]*length(s2sum));i1(1)=1;
[pksval,pksloc]=findpeaks(s2sum);
indr=(find(pksloc>=i1(4) & pksloc<=i1(6)));indl=(find(pksloc>=i1(1) & pksloc<=i1(3)));
% max(pksval(indr))
% max(pksval(indl))
if isempty(indl)
%     disp('r1');
    [v, id]=max(pksval(indr));
    s2loc=pksloc(indr(id))-round(avg/2);
elseif isempty(indr)
%     disp('l1');
    [v, id]=max(pksval(indl));
    s2loc=pksloc(indl(id))-round(avg/2);
else
    if max(pksval(indr))>max(pksval(indl))
%         disp('r');
        [v, id]=max(pksval(indr));
        s2loc=pksloc(indr(id))-round(avg/2);
    else
%         disp('l');
        [v, id]=max(pksval(indl));
        s2loc=pksloc(indl(id))-round(avg/2);
    end
end
% indr
% v
% id
% s2loc
% figure(101);plot(-round(avg/2):1:round(avg/2),s2sum);title(s2loc);
if s2loc>0
    indices_s1=indices;indices_s2=indices+s2loc;
    indices_s2(indices_s2>length(lp_res))=[];
    indices_s2(indices_s2<1)=[];
else
    indices_s2=indices;indices_s1=indices+s2loc;
    indices_s1(indices_s1>length(lp_res))=[];
    indices_s1(indices_s1<1)=[];
end
